### PR TITLE
feat(client): Show the  button on verification tab in Fennec

### LINKED ISSUE #3140: Enable the "Sync preferences" button from the verification page for Fennec
Currently we only show the "Sync preferences" button in the signup tab, but it would probably be more useful on the verification page because that's where the user is.

The "Sync preferences" button should be shown on the verification page if:
1) The user verifies in the same browser that they signed up in.
2) The same user is verifying as signed up.

Ref #2771

### PR DESCRIPTION
**Pull request by shane-tomlinson**
Save verification info into localStorage that says what broker was used
during sign up. If the user verifies in the same browser, use the same
broker.

fixes #3140

### CODE REVIEW

File: app/scripts/models/auth_brokers/base.js [hunk 1]
```diff
@@ -140,18 +146,36 @@ define([
      * Called before confirmation polls to persist any data that is needed
      * for email verification. Useful for storing data that may be needed
      * by the verification tab.
+     *
+     * @param {object} account
+     * @return {promise}
      */
-    persist: function () {
-      return p();
+    persist: function (account) {
```
Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/base.js]: pass in a `verificationInfo`  object to `persist`, it's fine to add `context` to that. Passing in the object will allow each broker to add data that it needs stored,

File: app/scripts/models/auth_brokers/base.js [hunk 2]
```diff
@@ -171,18 +196,20 @@ define([
      * to notify the RP that the user has successfully signed up and
      * completed verification.
      *
+     * @param {object} account
```
Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/base.js]: Do these doc updates in a separate PR.

File: app/scripts/lib/app-start.js [hunk 3]
```diff
@@ -706,7 +708,26 @@ function (
...
     _getContext: function () {
-      return this._searchParam('context') || Constants.DIRECT_CONTEXT;
+      if (this._isVerification()) {
+        return this._getVerificationContext();
+      } else {
+        return this._searchParam('context') || Constants.DIRECT_CONTEXT;
+      }
+    },
+
+    _getVerificationContext: function () {
```
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: Add a test for this function.

File: app/scripts/lib/app-start.js [hunk 4]
```diff
@@ -706,7 +708,26 @@ function (
...
+      // Users that verify in the same browser use the same context that
+      // was used to sign up to allow the verification tab to have
+      // the same capabilities as the signup tab.
+      // If verifying in a separate browser, fall back to the default context.
+      var verificationInfo = this._getVerficationInfo();
+      return verificationInfo.get('context') || Constants.DIRECT_CONTEXT;
+    },
+
+    _getVerficationInfo: function () {
```
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: Add a test for this function, and rename to `_getSameBrowserVerificationInfo` or similar.

File: app/scripts/models/verification/base.js [hunk 5]
```diff
@@ -35,11 +37,15 @@ define([
...
 
+    defaults: {},
+    validation: {},
+
     /**
      * Check if the model is valid.
      *
      * @method isValid
-     * @returns `false` if a `validationError` is set, or if `validate` either throws or returns false. `true` otherwise.
+     * @returns {boolean} `false` if a `validationError` is set, or if
```
Comment (shane-tomlinson) [inline app/scripts/models/verification/base.js]: Put the changes from this file into a separate PR.

File: app/scripts/views/complete_reset_password.js [hunk 6]
```diff
@@ -178,6 +180,13 @@ function (Cocktail, BaseView, FormView, Template, FloatingPlaceholderMixin,
         });
       })
       .fail(self.displayError.bind(self));
+    },
+
+    _clearSameBrowserVerificationInfo: function () {
```
Comment (shane-tomlinson) [inline app/scripts/views/complete_reset_password.js]: Maybe move this function into a mixin to be shared. Expose `urlVerificationInfo` as a public variable, or create a getter function.

File: app/scripts/models/verification/same-browser.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    },
+
+    /**
+     * Clear verification info from localStorage
+     */
+    clear: function () {
+      var id = this._getUsersStorageId();
+      if (id) {
+        var verificationInfo = this._storage.get(STORAGE_KEY) || {};
+        delete verificationInfo[id][this._namespace];
```
Comment (shane-tomlinson) [inline app/scripts/models/verification/same-browser.js]: make sure verificationInfo[id] exists before deleting its members.

**Comment by shane-tomlinson** [comment]
The failing functional tests are caused by https://github.com/mozilla/fxa-content-server/issues/3237

**Comment by shane-tomlinson** [comment]
I'm taking the WIP off. This is ready for review. @philbooth or @vladikoff - r?

File: app/scripts/models/auth_brokers/base.js [hunk 8]
```diff
@@ -193,8 +224,12 @@ define([
      * @param {object} account
      * @return {promise}
      */
-    afterCompleteSignUp: function (/* account */) {
-      return p(this.getBehavior('afterCompleteSignUp'));
+    afterCompleteSignUp: function (account) {
+      var self = this;
+      return self.clearPersisted(account)
```
Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/base.js]: In a soonish-future, I'd like to have the brokers manage `persist` as well, and move that responsibility [out of the views](https://github.com/mozilla/fxa-content-server/blob/master/app/scripts/views/confirm.js#L99)

File: app/scripts/models/auth_brokers/base.js [hunk 9]
```diff
@@ -146,10 +149,38 @@ define([
...
+    persist: function (account) {
+      var self = this;
+
+      return p().then(function () {
+        // verification info is persisted to localStorage so that
+        // the same `context` is used if the user verifies in the same browser.
+        // If the user verifies in a different browser, the
+        // default (direct access) context will be used.
+        var verificationInfo =
+              createSameBrowserVerificationInfo(account, 'context');
```
Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/base.js]: Each SameBrowserVerificationInfo object gets a namespace, I'd like to move the [oauth information that is saved](https://github.com/mozilla/fxa-content-server/blob/9136ab5ee091e33ba76ec185c2a7bee34a6bce4d/app/scripts/models/auth_brokers/oauth.js#L138) to use this object too, I imagine that info would go under the `oauth` namespace.

File: app/scripts/models/auth_brokers/base.js [hunk 10]
```diff
@@ -146,10 +149,38 @@ define([
...
+      });
+    },
+
+    /**
+     * Clear persisted verification data for the account
+     *
+     * @param {object} account
+     * @return {promise}
+     */
+    clearPersisted: function (account) {
```
Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/base.js]: `clearPersisted`, `unpersist`?

File: app/scripts/models/auth_brokers/redirect.js [hunk 11]
```diff
@@ -86,14 +87,17 @@ define([
...
 
-      return p().delay(100).then(function () {
-        if (self.isOriginalTab() || self.canVerificationRedirect()) {
-          return self.finishOAuthSignUpFlow(account)
-            .then(function () {
-              return new HaltBehavior();
-            });
-        }
-      });
+      return proto.afterCompleteSignUp.call(self, account)
```
Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/redirect.js]: I noticed several brokers do not call the prototype function to get the default behavior. This should be fixed in a follow up issue.

File: app/scripts/models/verification/same-browser.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+  'lib/storage'
+], function (Backbone, Storage) {
+  'use strict';
+
+  var STORAGE_KEY = 'verificationInfo';
+
+  var proto = Backbone.Model.prototype;
+
+  var SameBrowserVerificationInfo = Backbone.Model.extend({
+    initialize: function (attributes, options) {
```
Comment (shane-tomlinson) [inline app/scripts/models/verification/same-browser.js]: I just found out about the `attribute, options` paradigm used by backbone. We conflate the two everywhere, which leads to a horrible situation where non-data options end up in attributes. We should fix this in a follow up issue.

File: app/scripts/models/auth_brokers/base.js (hunk 10 above)
Comment (philbooth) [inline app/scripts/models/auth_brokers/base.js]: The natural antonym is `desist`, any good?

File: app/scripts/lib/app-start.js [hunk 13]
```diff
@@ -60,6 +60,7 @@ define([
   'models/auth_brokers/iframe',
   'models/unique-user-id',
   'models/user',
+  'models/verification/same-browser',
```
Comment (vladikoff) [inline app/scripts/lib/app-start.js]: nit: we just "git" rid of `same browser` in https://github.com/mozilla/fxa-content-server/commit/094bfdf2f619f14b9666f200d68f80ac573380b1

Is this something we want to rename here?

File: app/scripts/models/auth_brokers/base.js (hunk 10 above)
Comment (vladikoff) [inline app/scripts/models/auth_brokers/base.js]: Hm, this doesn't explain what it is clearing with that name. 
Is `unpersistVerificationData` better?

File: app/scripts/lib/app-start.js (hunk 13 above)
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: @vladikoff, @philbooth - any suggestions on a better name? The idea I'm trying to convey is:

"Verification info that is used if the user verifies in the same browser."

File: tests/functional/fx_fennec_v1_sign_up.js [hunk 14]
```diff
@@ -111,6 +111,19 @@ define([
           })
         .end()
 
+          // user can open sync preferences in new tab.
+        .then(FunctionalHelpers.noSuchBrowserNotification(self, 'fxaccounts:sync_preferences'))
+
+        // user should be able to open sync preferences
+        .findByCssSelector('#sync-preferences')
+          // user wants to open sync preferences.
+          .click()
```
Comment (vladikoff) [inline tests/functional/fx_fennec_v1_sign_up.js]: :+1:

File: app/scripts/lib/app-start.js (hunk 13 above)
Comment (vladikoff) [inline app/scripts/lib/app-start.js]: Eh we can stay with "Same browser", just as long everyone is on board.

other ideas: "original browser", "existing browser"

File: app/scripts/lib/app-start.js [hunk 15]
```diff
@@ -111,6 +112,7 @@ function (
   IframeAuthenticationBroker,
   UniqueUserId,
   User,
+  SameBrowserVerificationInfo,
```
Comment (vladikoff) [inline app/scripts/lib/app-start.js]: nit: `SameBrowserVerificationInfo` -> SameBrowserVerificationModel

**Comment by vladikoff** [comment]
works well in my testing, some nits, r+ with travis passing

File: app/scripts/lib/app-start.js (hunk 15 above)
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: :+1: I dig it.

File: app/scripts/lib/app-start.js (hunk 13 above)
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: One idea I was tossing around is a universal "verification info model". We have some info persisted in localStorage, some from query parameters, and some in the resume token. Can we combine all those sources of information into a single, unified interface?

File: app/scripts/models/auth_brokers/base.js (hunk 10 above)
Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/base.js]: @vladikoff - I think `unpersistVerificationData` is definitely clearer than `clearPersisted`, perhaps we should update `persist` to be `persistVerificationData` while we are digging around. Thoughts?

Comment (vladikoff) [inline app/scripts/models/auth_brokers/base.js]: Sure, sounds good!
